File: library/src/main/java/com/vorlonsoft/android/rate/DialogOptions.java
/*
 * Copyright 2017 - 2018 Vorlonsoft LLC
 *
 * Licensed under The MIT License (MIT)
 */

package com.vorlonsoft.android.rate;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;

import java.lang.ref.SoftReference;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StyleRes;

import static com.vorlonsoft.android.rate.Constants.Utils.TAG;
import static com.vorlonsoft.android.rate.DialogType.CLASSIC;

/**
 * <p>DialogOptions Class - dialog options class of the AndroidRate library.</p>
 *
 * @since    0.5.1
 * @version  1.2.1
 * @author   Alexander Savin
 * @author   Shintaro Katafuchi
 */

public final class DialogOptions {

    private boolean cancelable = false;
    /** <p>Whether the Rate Dialog should show the message.</p> */
    private boolean showMessage = true;

    private boolean showNegativeButton = true;

    private boolean showNeutralButton = true;

    private boolean showTitle = true;
    /** Default value is null. */
    private Boolean showDialogIcon = null;

    private int dialogIconResId = 0;
    /** <p>One of the dialog types defined by {@link DialogType.AnyDialogType}.</p> */
    @DialogType.AnyDialogType
    private int dialogType = DialogType.CLASSIC;

    private int textMessageResId = R.string.rate_dialog_message;

    private int textNegativeResId = R.string.rate_dialog_no;

    private int textNeutralResId = R.string.rate_dialog_cancel;

    private int textPositiveResId = R.string.rate_dialog_ok;

    private int textTitleResId = R.string.rate_dialog_title;

    private Integer themeResId = null;

    private String messageText = null;

    private String negativeText = null;

    private String neutralText = null;

    private String positiveText = null;

    private String titleText = null;

    private Drawable dialogIcon = null;

    private View view = null;

    private SoftReference<OnClickButtonListener> listener = null;

    DialogOptions() {
    }

    /**
     * <p>Returns whether the Rate Dialog should show the message.</p>
     *
     * @return true to show the message, false otherwise
     */
    @SuppressWarnings("WeakerAccess")
    public boolean shouldShowMessage() {
        return showMessage;
    }

    /**
     * <p>Sets whether the Rate Dialog should show the message.</p>
     *
     * @param showMessage true to show the message, false otherwise
     * @see AppRate#setShowMessage(boolean)
     */
    void setShowMessage(final boolean showMessage) {
        this.showMessage = showMessage;
    }

    /**
     * <p>Decides whether the Neutral button ("Remind Me Later") appears in the Rate Dialog or
     * not.</p>
     *
     * @return true to show the Neutral button, false otherwise
     */
    @SuppressWarnings("WeakerAccess")
    public boolean shouldShowNeutralButton() {
        return showNeutralButton;
    }

    /**
     * <p>Sets whether the Neutral button ("Remind Me Later") appears in the Rate Dialog or
     * not.</p>
     *
     * @param showNeutralButton default is true, true means to show the Neutral button
     * @see AppRate#setShowLaterButton(boolean)
     */
    void setShowNeutralButton(final boolean showNeutralButton) {
        this.showNeutralButton = showNeutralButton;
    }

    /**
     * <p>Decides if the Negative button appears in the Rate Dialog or not.</p>
     *
     * @return true to show the Negative button, false otherwise
     */
    @SuppressWarnings("WeakerAccess")
    public boolean shouldShowNegativeButton() {
        return showNegativeButton;
    }

    /**
     * <p>Sets if the Negative button appears in the Rate Dialog or not.</p>
     *
     * @param showNegativeButton default is true, true means to show the Negative button
     * @see AppRate#setShowNeverButton(boolean)
     */
    void setShowNegativeButton(final boolean showNegativeButton) {
        this.showNegativeButton = showNegativeButton;
    }

    /**
     * <p>Whether the Rate Dialog should show the title.</p>
     *
     * @return true to show the title, false otherwise
     */
    @SuppressWarnings("WeakerAccess")
    public boolean shouldShowTitle() {
        return showTitle;
    }

    /**
     * <p>Sets whether the Rate Dialog should show the title.</p>
     *
     * @param showTitle true to show the title, false otherwise
     * @see AppRate#setShowTitle(boolean)
     */
    void setShowTitle(final boolean showTitle) {
        this.showTitle = showTitle;
    }

    /**
     * <p>Whether the Rate Dialog should show the icon.</p>
     *
     * @return true to show the icon, false otherwise
     */
    @SuppressWarnings("WeakerAccess")
    public boolean shouldShowDialogIcon() {
        return (showDialogIcon == null) ? !(dialogType == CLASSIC) : showDialogIcon;
    }

    /**
     * <p>Sets whether the Rate Dialog should show the icon.</p>
     *
     * @param showDialogIcon true to show the icon, false otherwise, default values are false for
     *                       {@link DialogType#CLASSIC DialogType.CLASSIC} and true for
     *                       {@link DialogType#APPLE DialogType.APPLE} and
     *                       {@link DialogType#MODERN DialogType.MODERN}
     * @see AppRate#setShowDialogIcon(boolean)
     */
    void setShowDialogIcon(final boolean showDialogIcon) {
        this.showDialogIcon = showDialogIcon;
    }

    /**
     * <p>Whether the Rate Dialog is cancelable or not.</p>
     *
     * @return true if the Rate Dialog is cancelable, false otherwise
     */
    @SuppressWarnings("WeakerAccess")
    public boolean getCancelable() {
        return cancelable;
    }

    /**
     * <p>Sets whether the Rate Dialog is cancelable or not.</p>
     *
     * @param cancelable default is false, true will set the Rate Dialog cancelable
     * @see AppRate#setCancelable(boolean)
     */
    void setCancelable(final boolean cancelable) {
        this.cancelable = cancelable;
    }

    /**
     * <p>Returns the Rate Dialog type.</p>
     *
     * @return one of the values defined by {@link DialogType.AnyDialogType}
     */
    @SuppressWarnings({"WeakerAccess", "unused"})
    @DialogType.AnyDialogType
    public int getDialogType() {
        return dialogType;
    }

    /**
     * <p>Sets the Rate Dialog type.</p>
     *
     * @param dialogType one of the values defined by {@link DialogType.AnyDialogType}
     */
    void setDialogType(@DialogType.AnyDialogType final int dialogType) {
        this.dialogType = dialogType;
    }

    /**
     * <p>Returns the Rate Dialog icon resource ID.</p>
     *
     * @return the Rate Dialog icon resource ID
     */
    @SuppressWarnings("unused")
    int getDialogIconResId() {
        return dialogIconResId;
    }

    /**
     * <p>Sets the Rate Dialog icon resource ID.</p>
     *
     * @param dialogIconResId the Rate Dialog icon resource ID
     * @see AppRate#setDialogIcon(int)
     * @see AppRate##setShowDialogIcon(boolean)
     */
    void setDialogIconResId(final int dialogIconResId) {
        this.dialogIconResId = dialogIconResId;
    }

    /**
     * <p>Returns the Rate Dialog title text.</p>
     *
     * @return the Rate Dialog title text resource ID
     */
    @SuppressWarnings("unused")
    int getTitleResId() {
        return textTitleResId;
    }

    /**
     * <p>Sets the Rate Dialog title text.</p>
     *
     * @param textTitleResId the Rate Dialog title text resource ID
     * @see AppRate#setTitle(int)
     * @see AppRate#setShowTitle(boolean)
     */
    void setTitleResId(final int textTitleResId) {
        this.textTitleResId = textTitleResId;
    }

    @SuppressWarnings("unused")
    int getMessageResId() {
        return textMessageResId;
    }

    void setMessageResId(final int textMessageResId) {
        this.textMessageResId = textMessageResId;
    }

    @SuppressWarnings("unused")
    int getTextPositiveResId() {
        return textPositiveResId;
    }

    void setTextPositiveResId(final int textPositiveResId) {
        this.textPositiveResId = textPositiveResId;
    }

    @SuppressWarnings("unused")
    int getTextNeutralResId() {
        return textNeutralResId;
    }

    void setTextNeutralResId(final int textNeutralResId) {
        this.textNeutralResId = textNeutralResId;
    }

    @SuppressWarnings("unused")
    int getTextNegativeResId() {
        return textNegativeResId;
    }

    void setTextNegativeResId(final int textNegativeResId) {
        this.textNegativeResId = textNegativeResId;
    }

    /**
     * <p>Returns Rate Dialog View. <b>Use only for {@link DialogType#CLASSIC DialogType.CLASSIC}
     * dialogs!</b></p>
     *
     * @return Rate Dialog View
     * @deprecated since 1.2.2 use {@link #getView(Context)} instead
     * @see #getView(Context)
     */
    @SuppressWarnings("WeakerAccess")
    @Nullable
    @Deprecated
    public View getView() {
        if (dialogType != CLASSIC) {
            Log.w(TAG, "For other than DialogType.CLASSIC dialog types call " +
                       "DialogOptions#getView(Context) instead of DialogOptions#getView().");
        }
        return getView(null);
    }

    /**
     * <p>Returns Rate Dialog View.</p>
     *
     * @param dialogContext a Context with the appropriate theme for built dialogs
     * @return Rate Dialog View
     * @since 1.2.2
     */
    @SuppressLint("InflateParams")
    @SuppressWarnings("WeakerAccess")
    @Nullable
    public View getView(final Context dialogContext) {
        if ((view == null) && !(dialogType == CLASSIC)) {
            if (dialogContext != null) {
                try {
                    return LayoutInflater.from(dialogContext).inflate(R.layout.rate_dialog,
                                                                      null);
                } catch (AssertionError e) {
                    Log.i(TAG, "Can't inflate the R.layout.rate_dialog, layoutInflater not " +
                               "found, DialogType.CLASSIC dialog will be used.");
                }
            } else {
                Log.w(TAG, "Can't inflate the R.layout.rate_dialog, dialogContext == null, " +
                           "DialogType.CLASSIC dialog will be used.");
            }
        }
        return view;
    }

    void setView(@Nullable final View view) {
        this.view = view;
    }

    @Nullable
    OnClickButtonListener getListener() {
        return listener != null ? listener.get() : null;
    }

    void setListener(final OnClickButtonListener listener) {
        this.listener = new SoftReference<>(listener);
    }

    /**
     * <p>Returns the Rate Dialog title text.</p>
     *
     * @param context activity context
     * @return the Rate Dialog title text
     */
    @SuppressWarnings("WeakerAccess")
    public String getTitleText(final Context context) {
        if (titleText == null) {
            return context.getString(textTitleResId);
        }
        return titleText;
    }

    /**
     * <p>Sets the Rate Dialog title text.</p>
     *
     * @param titleText the Rate Dialog title text
     * @see AppRate#setTitle(int)
     * @see AppRate#setShowTitle(boolean)
     */
    void setTitleText(final String titleText) {
        this.titleText = titleText;
    }

    @SuppressWarnings("WeakerAccess")
    public String getMessageText(final Context context) {
        if (messageText == null) {
            return context.getString(textMessageResId);
        }
        return messageText;
    }

    void setMessageText(final String messageText) {
        this.messageText = messageText;
    }

    @SuppressWarnings("WeakerAccess")
    public String getPositiveText(final Context context) {
        if (positiveText == null) {
            return context.getString(textPositiveResId);
        }
        return positiveText;
    }

    void setPositiveText(final String positiveText) {
        this.positiveText = positiveText;
    }

    @SuppressWarnings("WeakerAccess")
    public String getNeutralText(final Context context) {
        if (neutralText == null) {
            return context.getString(textNeutralResId);
        }
        return neutralText;
    }

    void setNeutralText(final String neutralText) {
        this.neutralText = neutralText;
    }

    @SuppressWarnings("WeakerAccess")
    public String getNegativeText(final Context context) {
        if (negativeText == null) {
            return context.getString(textNegativeResId);
        }
        return negativeText;
    }

    void setNegativeText(final String negativeText) {
        this.negativeText = negativeText;
    }

    @SuppressWarnings("WeakerAccess")
    @Nullable
    public Drawable getDialogIcon(@NonNull final Context context) {
        if (dialogIconResId != 0) {
            try {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    return context.getDrawable(dialogIconResId);
                } else {
                    //noinspection deprecation
                    return context.getResources().getDrawable(dialogIconResId);
                }
            } catch (android.content.res.Resources.NotFoundException e) {
                Log.i(TAG, "Dialog icon with the given ResId doesn't exist.");
            }
        }
        return (dialogIcon != null) ? dialogIcon : AppInformation.getInstance(context).getAppIcon();
    }

    void setDialogIcon(final Drawable dialogIcon) {
        this.dialogIcon = dialogIcon;
    }

    @SuppressWarnings("WeakerAccess")
    @StyleRes
    public int getThemeResId() {
        if (themeResId == null) {
            return !(dialogType == CLASSIC) ? R.style.RateDialogTransparentTheme : 0;
        }
        return themeResId;
    }

    void setThemeResId(@StyleRes final int themeResId) {
        this.themeResId = themeResId;
    }
}

File: library/src/main/java/com/vorlonsoft/android/rate/DialogOptions.kt
/*
 * Copyright 2017 - 2018 Vorlonsoft LLC
 *
 * Licensed under The MIT License (MIT)
 */

package com.vorlonsoft.android.rate

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.drawable.Drawable
import android.os.Build.VERSION.SDK_INT
import android.os.Build.VERSION_CODES.LOLLIPOP
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import androidx.annotation.StyleRes
import com.vorlonsoft.android.rate.Constants.Utils.EMPTY_STRING
import com.vorlonsoft.android.rate.Constants.Utils.TAG
import com.vorlonsoft.android.rate.DialogType.Companion.CLASSIC
import java.lang.ref.SoftReference

/**
 * DialogOptions Class - dialog options class of the AndroidRate library.
 *
 * Contains dialog properties with setters and getters.
 *
 * @since    0.5.1
 * @version  1.2.1
 * @author   Alexander Savin
 * @author   Shintaro Katafuchi
 */
class DialogOptions internal constructor() {
    /** Whether the Rate Dialog is cancelable or not. */
    var cancelable: Boolean = false

    /** Whether the current Rate Dialog had redrawn one or more times. */
    var isRedrawn: Boolean = false

    /** Whether the Rate Dialog should show an icon. Default value is null. */
    var isShowIcon: Boolean? = null
        get() = if (field != null) field else (type != CLASSIC)

    /** Whether the Rate Dialog should show the message. */
    var isShowMessage: Boolean = true
    /**
     * Returns whether the Rate Dialog should show the message.
     *
     * @return true to show the message, false otherwise
     */
    @Suppress("unused")
    fun shouldShowMessage(): Boolean {
        return isShowMessage
    }

    /** Whether the Rate Dialog should show the Negative button. */
    var isShowNegativeButton: Boolean = true
    /**
     * Returns whether the Rate Dialog should show the Negative button.
     *
     * @return true to show the Negative button, false otherwise
     */
    @Suppress("unused")
    fun shouldShowNegativeButton(): Boolean {
        return isShowNegativeButton
    }

    /** Whether the Rate Dialog should show the Neutral button. */
    var isShowNeutralButton: Boolean = true
    /**
     * Returns whether the Rate Dialog should show the Neutral button.
     *
     * @return true to show the Neutral button, false otherwise
     */
    @Suppress("unused")
    fun shouldShowNeutralButton(): Boolean {
        return isShowNeutralButton
    }

    /** Whether the Rate Dialog should show the title. */
    var isShowTitle: Boolean = true
    /**
     * Returns whether the Rate Dialog should show the title.
     *
     * @return true to show the title, false otherwise
     */
    @Suppress("unused")
    fun shouldShowTitle(): Boolean {
        return isShowTitle
    }

    /** The current rating set by the app user, 0 means that the user hasn't set a rating yet. */
    var currentRating: Byte = 0

    /** The Rate Dialog icon resource ID. */
    var iconResId: Int = 0

    /** The Rate Dialog message text resource ID. */
    var messageTextResId: Int = R.string.rate_dialog_message

    /** The Rate Dialog negative text resource ID. */
    var negativeTextResId: Int = R.string.rate_dialog_no

    /** The Rate Dialog neutral text resource ID. */
    var neutralTextResId: Int = R.string.rate_dialog_cancel

    /** The Rate Dialog positive text resource ID. */
    var positiveTextResId: Int = R.string.rate_dialog_ok

    /** The Rate Dialog theme resource ID. */
    var themeResId: Int? = null
        @StyleRes get() = if (field != null) field else
            if (type != CLASSIC) R.style.RateDialogTransparentTheme else 0

    /** The Rate Dialog title text resource ID. */
    var titleTextResId: Int = R.string.rate_dialog_title

    /** The Rate Dialog type. One of the values defined by [DialogType.AnyDialogType]. */
    @DialogType.AnyDialogType
    var type: Int = DialogType.CLASSIC
        @DialogType.AnyDialogType get
    /**
     * Returns the Rate Dialog type.
     *
     * @return one of the values defined by [DialogType.AnyDialogType]
     */
    @Suppress("unused")
    @DialogType.AnyDialogType
    fun getDialogType(): Int {
        return type
    }

    /** The Rate Dialog message text. */
    private var messageText: String? = null
    /**
     * Creates the Rate Dialog message text.
     *
     * @param context activity context
     * @return the Rate Dialog message text
     */
    fun getMessageText(context: Context): String {
        return if (messageText == null) context.getString(messageTextResId)
        else messageText ?: EMPTY_STRING
    }
    /**
     * Sets the Rate Dialog message text.
     *
     * @param messageText the Rate Dialog message text
     */
    fun setMessageText(messageText: String) {
        this.messageText = messageText
    }

    /** The Rate Dialog negative text. */
    private var negativeText: String? = null
    /**
     * Creates the Rate Dialog negative text.
     *
     * @param context activity context
     * @return the Rate Dialog negative text
     */
    fun getNegativeText(context: Context): String {
        return if (negativeText == null) context.getString(negativeTextResId)
        else negativeText ?: EMPTY_STRING
    }
    /**
     * Sets the Rate Dialog negative text.
     *
     * @param negativeText the Rate Dialog negative text
     */
    fun setNegativeText(negativeText: String) {
        this.negativeText = negativeText
    }

    /** The Rate Dialog neutral text. */
    private var neutralText: String? = null
    /**
     * Creates the Rate Dialog neutral text.
     *
     * @param context activity context
     * @return the Rate Dialog neutral text
     */
    fun getNeutralText(context: Context): String {
        return if (neutralText == null) context.getString(neutralTextResId)
        else neutralText ?: EMPTY_STRING
    }
    /**
     * Sets the Rate Dialog neutral text.
     *
     * @param neutralText the Rate Dialog neutral text
     */
    fun setNeutralText(neutralText: String) {
        this.neutralText = neutralText
    }

    /** The Rate Dialog positive text. */
    private var positiveText: String? = null
    /**
     * Creates the Rate Dialog positive text.
     *
     * @param context activity context
     * @return the Rate Dialog positive text
     */
    fun getPositiveText(context: Context): String {
        return if (positiveText == null) context.getString(positiveTextResId)
        else positiveText ?: EMPTY_STRING
    }
    /**
     * Sets the Rate Dialog positive text.
     *
     * @param positiveText the Rate Dialog positive text
     */
    fun setPositiveText(positiveText: String) {
        this.positiveText = positiveText
    }

    /** The Rate Dialog title text. */
    private var titleText: String? = null
    /**
     * Creates the Rate Dialog title text.
     *
     * @param context activity context
     * @return the Rate Dialog title text
     */
    fun getTitleText(context: Context): String {
        return if (titleText == null) context.getString(titleTextResId)
        else titleText ?: EMPTY_STRING
    }
    /**
     * Sets the Rate Dialog title text.
     *
     * @param titleText the Rate Dialog title text
     */
    fun setTitleText(titleText: String) {
        this.titleText = titleText
    }

    /** The Rate Dialog SoftReference to the buttons listener implemented by a library user. */
    private var buttonListener: SoftReference<OnClickButtonListener>? = null
    /**
     * Returns the Rate Dialog buttons on-click listener implemented by a library user.
     *
     * @return the Rate Dialog buttons on-click listener
     */
    fun getButtonListener(): OnClickButtonListener? {
        return buttonListener?.get()
    }
    /**
     * Sets the Rate Dialog SoftReference to the buttons on-click listener implemented by a library
     * user.
     *
     * @param buttonListener the Rate Dialog buttons on-click listener
     */
    fun setButtonListener(buttonListener: OnClickButtonListener) {
        this.buttonListener = SoftReference(buttonListener)
    }

    /** The Rate Dialog icon. */
    private var icon: Drawable? = null
    /**
     * Returns the Rate Dialog icon.
     *
     * @param context activity context
     * @return the Rate Dialog icon
     */
    fun getIcon(context: Context): Drawable? {
        return when {
            (iconResId != 0) -> try {
                if (SDK_INT >= LOLLIPOP) {
                    context.getDrawable(iconResId)
                } else {
                    @Suppress("DEPRECATION")
                    context.resources.getDrawable(iconResId)
                }
            } catch (e: android.content.res.Resources.NotFoundException) {
                Log.i(TAG, "Dialog icon with the given ResId doesn't exist.")
                if (icon != null) icon else AppInformation.getIcon(context)
            }
            (icon != null) -> icon
            else -> AppInformation.getIcon(context)
        }
    }
    /**
     * Sets the Rate Dialog icon.
     *
     * @param icon the Rate Dialog icon
     */
    fun setIcon(icon: Drawable?) {
        this.icon = icon
    }
    /**
     * Returns the Rate Dialog icon.
     *
     * @param context activity context
     * @return the Rate Dialog icon
     */
    @Suppress("unused")
    fun getDialogIcon(context: Context): Drawable? {
        return getIcon(context)
    }

    /** The Rate Dialog view. */
    private var view: View? = null
    /**
     * Returns the Rate Dialog view.
     *
     * @return the Rate Dialog view
     */
    @Deprecated("Since 1.2.5 use DialogOptions.getView(Context) instead.")
    fun getView(): View? {
        if (type != CLASSIC) {
            Log.w(TAG, "For other than DialogType.CLASSIC dialog types call the " +
                           "DialogOptions.getView(Context) instead of the DialogOptions.getView().")
        }
        return getView(null)
    }
    /**
     * Returns the Rate Dialog view.
     *
     * @param dialogContext a Context with an appropriate theme for built dialogs
     * @return the Rate Dialog view
     * @since 1.2.5
     */
    @SuppressLint("InflateParams")
    fun getView(dialogContext: Context?): View? {
        return if ((view == null) && (type != CLASSIC)) {
            var dialogView: View? = null
            if (dialogContext != null) try {
                dialogView = LayoutInflater.from(dialogContext).inflate(R.layout.rate_dialog, null)
            } catch (e: AssertionError) {
                Log.i(TAG, "Can't inflate the R.layout.rate_dialog, layoutInflater not found, " +
                                                          "DialogType.CLASSIC dialog will be used.")
            } else Log.w(TAG, "Can't inflate the R.layout.rate_dialog, dialogContext == null, " +
                                                          "DialogType.CLASSIC dialog will be used.")
            dialogView
        } else view
    }
    /**
     * Sets the Rate Dialog view.
     *
     * @param view the Rate Dialog view
     */
    fun setView(view: View?) {
        this.view = view
    }
}